akka-docs-dev/rst/scala/stream-customize.rst

```diff
@@ -129,10 +135,144 @@ Finally, there are two methods available for convenience to complete the stage a
+we return the “ball” to our upstream so that we get the new element. This is achieved by modifying the map
+example by adding a conditional in the ``onPush`` handler and decide between a ``pull(in)`` or ``push(out)`` call
+(and of course not having a mapping ``f`` function).
+
+.. includecode:: code/docs/stream/GraphStageDocSpec.scala#many-to-one
+
+To complete the picture we define a one-to-many transformation as the next step. We chose a straightforward example stage
+that emits every upstream element twice downstream. The conceptual wiring of this stage looks like this:
+
+|
+
+.. image:: ../images/graph_stage_duplicate.png
+   :align: center
+   :width: 300
+
+|
+
+This is a stage that has state: an option with the last element it has seen indicating if it
+has duplicated this last element already or not. We must also make sure to emit the extra element
+if the upstream completes.
+
+.. includecode:: code/docs/stream/GraphStageDocSpec.scala#one-to-many
+
+In this case a pull from downstream might be consumed by the stage itself rather
+than passed along upstream as the stage might contain an element it wants to
+push. Note that we also need to handle the case where the upstream closes while
+the stage still has elements it wants to push downstream. This is done by
+overriding `onUpstreamFinish` in the `InHandler` and provide custom logic
+that should happen when the upstream has been finished.
+
+This example can be simplified by replacing the usage of a mutable state with calls to
+``emitMultiple`` which will replace the handlers, emit each of multiple elements and then
+reinstate the original handlers:
+
+.. includecode:: code/docs/stream/GraphStageDocSpec.scala#simpler-one-to-many
+
+
+Finally, to demonstrate all of the stages above, we put them together into a processing chain,
+which conceptually would correspond to the following structure:
+
+
+|
+
+.. image:: ../images/graph_stage_chain.png
+   :align: center
+   :width: 700
+
+|
+
+In code this is only a few lines, using the ``via`` use our custom stages in a stream:
+
+.. includecode:: code/docs/stream/GraphStageDocSpec.scala#graph-stage-chain
+
+If we attempt to draw the sequence of events, it shows that there is one "event token"
+in circulation in a potential chain of stages, just like our conceptual "railroad tracks" representation predicts.
+
+
+|
+
+.. image:: ../images/graph_stage_tracks_1.png
+   :align: center
+   :width: 700
+
+|
+
 Completion
 ----------
 
-**This section is a stub and will be extended in the next release**
+Completion handling usually (but not exclusively) comes into the picture when processing stages need to emit
+a few more elements after their upstream source has been completed. We have seen an example of this in our
+first :class:`Duplicator` implementation where the last element needs to be doubled even after the upstream neighbor
+stage has been completed. This can be done by overriding the ``onUpstreamFinish`` method in ``InHandler``.
```

Comment:
I think it is, but please check, and document that it is safe :) (I am not sure for example that explicitly closing the output port does not interrupt the emitting)